module/mempool/consensus/exec_fork_suppressor.go

```diff
@@ -0,0 +1,345 @@
+	sealID := irSeal.ID()
+	blockID := irSeal.Seal.BlockID
+	log := s.log.With().
+		Hex("seal_id", sealID[:]).
+		Hex("block_id", blockID[:]).
+		Logger()
+
+	// CAUTION: potential edge case:
+	// Upon adding a new seal, the ejector of the wrapped mempool decides to eject the element which was just added.
+	// In this case, the ejected seal is _not_ the secondary index.
+	//  (a) we don't have any seals for the respective block stored in the secondary index (yet).
+	//  (b) the secondary index contains only one seal for the block, which
+	//      is different than the seal just ejected
+	set, found := s.sealsForBlock[irSeal.Seal.BlockID]
+	if !found { // case (a)
+		return
+	}
+	delete(set, irSeal.ID())
+	if len(set) == 0 {
+		delete(s.sealsForBlock, irSeal.Seal.BlockID)
+	}
+	log.Debug().Msg("ejected seal")
+}
+
+// Add adds the given seal to the mempool. Return value indicates whether or not seal was added to mempool.
+// Error returns:
+//   * engine.InvalidInputError (sentinel error)
+//     In case a seal fails one of the required consistency checks;
+func (s *ExecForkSuppressor) Add(newSeal *flow.IncorporatedResultSeal) (bool, error) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	if s.execForkDetected {
+		return false, nil
+	}
+
+	// STEP 1: ensure locally that newSeal's start and end state are non-empty values
+	// This wrapper is a temporary safety layer; we check all conditions that are
+	// required for its correct functioning locally, to not delegate safety-critical
+	// implementation aspects to external components
+	err := s.enforceValidStates(newSeal)
+	if err != nil {
+		return false, fmt.Errorf("invalid candidate seal: %w", err)
+	}
+	blockID := newSeal.Seal.BlockID
+
+	// STEP 2: enforce that newSeal's state transition does not conflict with other stored seals for the same block
+	otherSeals, found := s.sealsForBlock[blockID]
+	if found {
+		// already other seal for this block in mempool => compare consistency of results' state transitions
+		otherSeal := getArbitraryElement(otherSeals) // cannot be nil, as otherSeals is guaranteed to always contain at least one element
+		err := s.enforceConsistentStateTransitions(newSeal, otherSeal)
+		if errors.Is(err, ExecutionForkErr) {
+			return false, nil
+		}
+		if err != nil {
+			return false, fmt.Errorf("state consistency check failed: %w", err)
+		}
+	} // no conflicting state transition for this block known
+
+	// STEP 3: add newSeal to the wrapped mempool
+	added, err := s.seals.Add(newSeal) // internally de-duplicates
+	if err != nil {
+		return added, fmt.Errorf("failed to add seal to wrapped mempool: %w", err)
+	}
+	if !added { // if underlying mempool did not accept the seal => nothing to do anymore
+		return false, nil
+	}
+
+	// STEP 4: check whether wrapped mempool ejected the newSeal right away;
+	// important to prevent memory leak
+	newSealID := newSeal.ID()
+	if _, exists := s.seals.ByID(newSealID); !exists {
+		return added, nil
```

Comment:
Totally agree with your thinking and would prefer a different return value in this particular case. Before we decide what the return value should be, lets take a look at the larger picture:
 * `ExecForkSuppressor` is a wrapper a around `mempool.IncorporatedResultSeals`. So how does the implementation behave when you `Add` an entity and the mempool overflows?
   - `mempool.IncorporatedResultSeals` is implemented by   https://github.com/onflow/flow-go/blob/d14a1e7a634d3d458a09406f2bbff96bb0701777/module/mempool/stdmap/incorporated_result_seals.go#L21-L24 which calls into `Backend` https://github.com/onflow/flow-go/blob/d14a1e7a634d3d458a09406f2bbff96bb0701777/module/mempool/stdmap/backend.go#L131-L138 and subsequently into `Backdata` https://github.com/onflow/flow-go/blob/d14a1e7a634d3d458a09406f2bbff96bb0701777/module/mempool/stdmap/backend.go#L31-L40
   - Specifically, the `Backend` always stores the element in `Backdata`. The **return value indicates**
      - `false`: entity is already present in the mempool and was deduplicated 
      - `true`: entity is new from the mempool's perspective and can be added (ignoring mempool's capacity limitations) 
   - The return value makes **no assertion about whether or not an element is subsequently ejected** 
* **Essentially, `IncorporatedResultSeals.Add(seal)` returns `true` if the entity is addable. Even if the mempool overflows and decides to eject the  `seal` right away, the seal was still "addable".** Note that this convention applies to all mempools backed by `Backend`. 

**`ExecForkSuppressor` is only a wrapper, and higher-level business logic should be agnostic to whether or not the wrapper exists. Therefore, I concluded that the wrapper should implement the same convention as the wrapped mempool.**

Nevertheless, @arrivets you are raising a valid point: 
* Given that a mempool has the ability to eject elements, can we base some algorithmic decisions on the return value? 
* If the mempool returns `false`, it is telling us that the entity is already stored. However, this might change any moment, as another routine might concurrently add another entity leading to the ejection of our entity. 
* Likewise, if the mempool returns `true`, the entity might be ejected right away as well.